### DISCUSSION TITLE
Cache aware streaming ASR - discrepancies between training and inference

### DISCUSSION THREAD

**Discussion by tobygodwin** [NVIDIA-NeMo/NeMo#7010] [Q&A] [answered]
Hello,

Firstly thanks for putting together a great tool.

I've trained a cache aware streaming fastconformer model using the [standard config ](https://github.com/NVIDIA/NeMo/blob/main/examples/asr/conf/fastconformer/cache_aware_streaming/fastconformer_ctc_bpe_streaming.yaml) and the standard ctc asr [training script ](https://github.com/NVIDIA/NeMo/blob/main/examples/asr/asr_ctc/speech_to_text_ctc_bpe.py).

The model trains well, but when I go to evaluate the streaming performance using the [cache aware streaming inference script ](https://github.com/NVIDIA/NeMo/blob/main/examples/asr/asr_cache_aware_streaming/speech_to_text_cache_aware_streaming_infer.py) there is a drastic difference between offline and streaming performance (13% and 70% respectively), which surprised me because I thought that I had limited the attention context using `chunked_limited` during training, and that if this was matched during inference then equivalent performance would be achieved.  When I increase the --chunk_size during streaming inference, performance improves significantly to 19% which implies to me that the original model was trained without any restrictions on attention context.

I've confirmed that the default streaming configs are the same between training and inference (the standard parameters returned by `setup_streaming_params()` for my config). One thing I did notice is that `cache_last_channel` and `cache_last_time` are not being passed during the training forward pass, but they are during the streaming inference forward pass.  I guess this results in a mismatch between training and inference which would lead to differences in performance.

Am I missing something?  I would have thought that `cache_last_channel` etc should be included in the training procedures.

Any guidance would be much appreciated.

Many thanks

**Comment by VahidooX** [comment]
Which branch or nemo version are you using?
Caches are not need during training. They are being used during inference as whole input is not given at once. You should get the exact same accuracy with the streaming script. We have made some changes recently. Let me check everything and get back to you.

**Comment by tobygodwin** [comment]
thanks for your response.

I'm using the master branch on this commit:

52d3af7aa728803aa0d10b66521bcc41d26f0db2

**Comment by VahidooX** [comment, marked as answer]
I found a bug in updating the caches and fixed it in this PR:
https://github.com/NVIDIA/NeMo/pull/7034

Would you please try it out?

**Comment by tobygodwin** [comment]
Sorry for the delay. That worked perfectly! Thanks very much

**Comment by sangeet2020** [comment]
Hello @titu1994 ,

I am in the process of training this model for German language, and I obtained these training stats after 100 epochs.

![image](https://github.com/NVIDIA/NeMo/assets/33633418/e7241292-d437-407b-b888-6b90990474a0)

Looking at the training loss and train wer, I believe the losses have not yet converged. Do you think something is wrong with the parameters/hyper-params. or does the training needs to continue?


Thank You

**Reply by titu1994** [reply]
Please don't comment on old discussions, open a new issue, since your discussion didn't seem related to this one. 

In any case, it looks like youre convergence is slow. Try using a pretrained model to speed it up. If it's chunk aware specifically, then training longer will help)

**Reply by sangeet2020** [reply]
Apologies. Will do so. thank you
